File: sdk/src/types/commission.ts
import { PublicKey } from "@solana/web3.js";
import BN from "bn.js";

export type CommissionStatus =
  | "open"
  | "matched"
  | "completed"
  | "cancelled"
  | "expired";

export interface Commission {
  commissionId: number;
  delegator: PublicKey;
  taskType: string;
  taskSpecHash: number[];
  taskSpecUri: string;
  maxPrice: BN;
  deadline: BN;
  status: CommissionStatus;
  selectedExecutor: PublicKey | null;
  selectedBidPrice: BN | null;
  bidCount: number;
  createdAt: BN;
  matchedAt: BN | null;
  completedAt: BN | null;
  minExecutorTier: number | null;
  bump: number;
  /** The on-chain PDA address */
  address: PublicKey;
}

/** Parse the Anchor enum format { open: {} } → "open" */
export function parseCommissionStatus(raw: any): CommissionStatus {
  if (raw.open !== undefined) return "open";
  if (raw.matched !== undefined) return "matched";
  if (raw.completed !== undefined) return "completed";
  if (raw.cancelled !== undefined) return "cancelled";
  if (raw.expired !== undefined) return "expired";
  throw new Error(`Unknown commission status: ${JSON.stringify(raw)}`);
}

export interface CreateCommissionParams {
  taskType: string;
  /** Full task spec JSON — will be SHA-256 hashed for on-chain verification */
  taskSpec: Record<string, any>;
  /** Short URI pointing to the full task spec (max 128 chars). e.g. IPFS, HTTP, or arweave URL */
  taskSpecUri: string;
  maxPrice: number;
  deadline: string | number;
  /** Optional minimum executor tier (0=Guest, 1=Trusted, 2=Verified, 3=Elite) */
  minExecutorTier?: number;
}

File: programs/inter-knot/src/state/commission.rs
use anchor_lang::prelude::*;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace)]
#[repr(u8)]
pub enum CommissionStatus {
    Open = 0,
    Matched = 1,
    Completed = 2,
    Cancelled = 3,
    Expired = 4,
}

#[account]
#[derive(InitSpace)]
pub struct Commission {
    /// Unique ID (from config.commission_count)
    pub commission_id: u64,
    /// Creator's wallet
    pub delegator: Pubkey,
    /// e.g. "compute/llm-inference"
    #[max_len(32)]
    pub task_type: String,
    /// SHA-256 of the full task_spec JSON
    pub task_spec_hash: [u8; 32],
    /// URI pointing to full task_spec JSON
    #[max_len(128)]
    pub task_spec_uri: String,
    /// Maximum price in USDC (6 decimals, so 1_000_000 = 1 USDC)
    pub max_price: u64,
    /// Bidding deadline (Unix timestamp)
    pub deadline: i64,
    /// Commission status
    pub status: CommissionStatus,
    /// Winning executor's wallet (set on match)
    pub selected_executor: Option<Pubkey>,
    /// Winning bid price (set on match)
    pub selected_bid_price: Option<u64>,
    /// Number of active bids
    pub bid_count: u32,
    /// Creation timestamp
    pub created_at: i64,
    /// When match was made
    pub matched_at: Option<i64>,
    /// When marked complete
    pub completed_at: Option<i64>,
    /// Minimum executor reputation tier required to bid (0=Guest,1=Trusted,2=Verified,3=Elite)
    pub min_executor_tier: Option<u8>,
    /// PDA bump seed
    pub bump: u8,
}
